### PR TITLE
Refactor applyPercentageDiscount() in CartRow class

### PR DESCRIPTION
**Pull request by arouiadib**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       |  develop
| Description?  | This PR tries to optimize some few lines of code by rewriting `applyPercentageDiscount()` `in CartRow` class, allowing code reuse  in the processing of cheapest product.
| Type?         |  refacto
| Category?     |  CO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes nothing.
| How to test?  | Nothing to test.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17686)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by SD1982** [comment]
Ping @atomiix , @Progi1984  How can we test this PR ?

**Comment by atomiix** [comment]
@SD1982 I think you can try to add a discount in Catalog > Discount > Add new cart rule. In the action tab, add a discount (in %) on the cheapest product. Then on the FO, try different scenarios on the cart and see if amounts are correct.

**Comment by SD1982** [comment]
Hello @arouiadib  
there is one error in this PR:

Refactoring applyPercentageDiscount() in CartRow class #17687 after PR


I applied a reduction of 50% on the cheapest article (1) whose price (2) is 14.28 € the reduction applied (3) is therefore 7.14 €. However in (4) the sum is calculated on the amount of the basket which is not good. The overall calculation is good, it's just the display in (4) which is incorrect. Can you review this point on your PR ?

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
It's related to the CartRuleCalculator, it needs behat tests.

**Comment by arouiadib** [comment]
Thank you very much @SD1982 for your time. 

The issue you are mentioning is a separate issue: ["The promo value is not correct in the cart summary & checkout summary"](https://github.com/PrestaShop/PrestaShop/issues/16842). I have created [ a PR ](https://github.com/PrestaShop/PrestaShop/pull/17502) for that issue.

This PR adds no functionality and it breaks no current tests. There is no real way of testing it. It just rewrite some few lines so that code is reused. The rewritten function has very limited and restrained dependency with other code.( But I am deeply in writing some integration tests,  WIP)

Sorry if I am long. I thank you very much guys for your quick feedbacks, time and support.

**Comment by PierreRambaud** [comment]
Hi @arouiadib 
Is this PR still in Work In Progress?

**Comment by arouiadib** [comment]
Hi @PierreRambaud ,apparently not.  But it is an old PR. I need  to check today  its status again.
I will be feedbacking. Thanks

File: src/Core/Cart/CartRow.php [hunk 1]
```diff
@@ -447,17 +447,19 @@ public function applyFlatDiscount(AmountImmutable $amount)
 
     /**
      * @param float $percent 0-100
+     * @param bool $onUnitPrice
      *
      * @return AmountImmutable
      */
-    public function applyPercentageDiscount($percent)
+    public function applyPercentageDiscount($percent, $onUnitPrice = false)
```
Comment (Progi1984) [inline src/Core/Cart/CartRow.php]:
```suggestion
    public function applyPercentageDiscount($percent, bool $onUnitPrice = false)
```

File: src/Core/Cart/CartRow.php [hunk 2]
```diff
@@ -447,17 +447,19 @@ public function applyFlatDiscount(AmountImmutable $amount)
...
     {
         $percent = (float) $percent;
         if ($percent < 0 || $percent > 100) {
             throw new \Exception('Invalid percentage discount given: ' . $percent);
         }
-        $discountTaxIncluded = $this->finalTotalPrice->getTaxIncluded() * $percent / 100;
-        $discountTaxExcluded = $this->finalTotalPrice->getTaxExcluded() * $percent / 100;
+        $priceUsed = (bool) $onUnitPrice ? $this->initialUnitPrice : $this->finalTotalPrice;
```
Comment (Progi1984) [inline src/Core/Cart/CartRow.php]:
```suggestion
        $priceUsed = $onUnitPrice ? $this->initialUnitPrice : $this->finalTotalPrice;
```

**Review by Progi1984 — COMMENTED** [review]
Two small feedbacks but it's good job.

**Review by Progi1984 — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Hi @arouiadib ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.
As I said above, since it touches the CartCalculator system it requires unit tests and/or integration tests.

Thanks!

**Pull request closed without merging**